Naga/c plusplus

naga-karupi opened the pull request.

naga-karupi committed:
section6 witten date and name

naga-karupi committed:
Merge branch 'main' of https://github.com/TNCTRobocon/manuals_markdown into naga/c_plusplus

naga-karupi committed:
Merge branch 'main' of https://github.com/TNCTRobocon/manuals_markdown into naga/c_plusplus

naga-karupi committed:
add sention7_type_inference.md

naga-karupi commented:
コメントテスト

naga-karupi committed:
add decltype and postscript auto type in section7

naga-karupi committed:
section8_new_delete.md

naga-karupi committed:
postscript comcept, typename and using in section7 and new operator calls constructor in section8

naga-karupi committed:
section9.1_class_introdection

naga-karupi committed:
fix words section9.1

naga-karupi committed:
fix word 'drow'

naga-karupi committed:
add section9.2_class_constructor.md
delete "new operator calls constructor" in section8_new_delete.md
instead, it written in section9.2_class_constructor.md

naga-karupi committed:
add column1_const.md
fixing section9.2_class_constructor.md now

naga-karupi committed:
add section9.3_class_operator.md
fix section9.1 and section9.2
postscript const method in  column1

naga-karupi committed:
add section9.4_class_static.md
fix column1_const.md, section9.3_class_operator.md

terurin commented on C++/C++/column1_const.md:
```diff
@@ -0,0 +1,184 @@
+
+また、返り値に`const`定義のされていない参照変数をメンバとして返すとエラーになる  
+まぁ`const`定義して返り値に指定するとエラーは出ない
+
```
mutableというクラスの属性に一応触れておくべき。

terurin commented on C++/C++/section8_new_delete.md:
```diff
@@ -0,0 +1,66 @@
+`new`や`delete`はほぼほぼ`malloc`や`free`の代用として利用できる  
+しかし、`malloc`との互換性があるわけではない  
+`new`で確保したメモリは`free()`で解放できないし、`malloc`で確保したメモリは`delete`で解放もできない
+
```
二重にdeleteもしくはfreeしないための慣例として開放したら速やかにnullptrを代入する作法がある。

terurin commented on C++/C++/section9.1_class_introduction.md:
```diff
@@ -0,0 +1,208 @@
+アクセス修飾子を書いた先からメンバ変数はそのアクセス制限の対象になる  
+あたらしい修飾子を書くとそこからはまた別のアクセス制限を受ける  
+アクセス修飾子をつけない場合は自動的にprivateを指定したとみなされる
+
```
structの場合、デフォルトでpublicになる

terurin commented on C++/C++/section9.1_class_introduction.md:
```diff
@@ -0,0 +1,208 @@
+
+メンバ変数がprivateになると`instance.member`といった方法でアクセスできない変数や関数も出てくる  
+その時にメンバ変数をpublicなメンバ関数を読み書きするためのメンバ関数を**アクセサ**といい、値を取得するアクセッサを**getter**、値を設定するためのものを**setter**という
+
```
セッター、ゲッターは次の観点から推奨できない。

1.カプセル化が破壊される
2.値を変える意図が分かりづらくなる。→ふさわしいメソッド名がないか検討するべき

terurin commented on C++/C++/section9.2_class_constructor.md:
```diff
@@ -0,0 +1,370 @@
+
+privateでコピーコンストラクタを定義するとインスタンスを生成した時点でエラーになる  
+
+自動生成されるコンストラクタは別の記法でも禁止することができる(C++17以降)
```
11からでは?

https://cpprefjp.github.io/lang/cpp11/defaulted_and_deleted_functions.html

terurin commented on C++/C++/section9.3_class_operator.md:
```diff
@@ -0,0 +1,86 @@
+`this->(変数やメソッド)`といったように自身のメンバにアクセスできる
+
+かなり見やすくなるので重宝する
+
```
技術的な点として、メソッドを呼び出す際に
暗黙的に第一引数としてthisが渡されていることを知っておくと良い。

terurin commented on C++/C++/section9.4_class_static.md:
````diff
@@ -0,0 +1,66 @@
+}
+```
+
+このようにインスタンスs1を通してs2の、s2を通してs1の静的メンバ変数`static_num`が書き変わっているのがわかるだろうか
````
書き換えが行われる静的変数は、マルチスレッドプログラミングにおいて原子性を保つことが困難であることから避けられる傾向にある。

terurin replied:
↑同様の理由でデザインパターンの一つであったシングルトンも避けられている

naga-karupi committed:
add exercise.md
add section10.1_class_inheritance.md

rewrite and postscript section8_new_delete.md, section9

terurin commented on C++/C++/exercise.md:
```diff
@@ -0,0 +1,49 @@
+初級 : 2×2行列または3×3行列の場合のプログラムを書け
+
+上級 : 任意の大きさの行列の積を計算できるプログラムを作れ  
+また、掛け算をできないときは-1と出力せよ
```
c++17以降ならstd optionalがオススメです。
あと行列同士または行列とベクトル、行列とスカラーいずれもスカラーになることを保証できないのでは?

terurin commented on C++/C++/section8_new_delete.md:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    return i;
+}
```
動かないぞい☆

terurin commented on C++/C++/section9.4_class_static.md:
```diff
@@ -0,0 +1,72 @@
+
+クラスに依存しているので`クラス名::静的メンバ`という風にアクセスすることも可能である
+
+ただし、書き換え可能な静的メンバ変数避けたほうが良い
```
は

terurin replied:
日本語を直そう

naga-karupi committed:
postscript exercise.md,
           section8_new_delete.md,
           section9.4_class_static.md and
           section10.1_class_inheritance.md

terurin commented on C++/C++/column1_const.md:
```diff
@@ -0,0 +1,186 @@
+置換が行われる範囲が極端に広くなる、すると……  
+怖いねー  
+**絶対にやめてください**
+
```
仮に使う場合もソースファイルで次のようにマクロが有効な範囲を最小化して利用しよう。

```c
#define HOGE(x) printf(hoge)
HOGE("abc");
#undef HOGE
```

terurin replied:
また、機種・基板ごとに処理を変更するときに利用できる`#ifdef`, `#ifndef`等は比較的有益である。
グローバルなマクロの設定は、makefile,cmake, プロジェクトの設定から行える。

terurin commented on C++/C++/section10.1_class_inheritance.md:
```diff
@@ -0,0 +1,129 @@
+継承より依存度が低いので含むクラスの設計が変更されても継承よりは影響が低いという関係性がある
+
+## is-a関係とhas-a関係
+
```
保守性の観点から、has-aの関係を駆使するほうが望ましい。

naga-karupi committed:
add section10.2_class_virtual.md
postscript column1_const, section10.1_class_inheritance

naga-karupi committed:
add section10.3_abstruct_class.md
postscript section10.2_class_virtual.md

naga-karupi committed:
add section10.4_inheritance_maltiple.md and pic
postscript section10.3_abstruct_class.md

terurin commented on C++/C++/column1_const.md:
````diff
@@ -0,0 +1,197 @@
+
+```c++
+const int const_i;
+const_i = 0;//ok
````
`const XX xx;`のようなコードは初期化していないとエラーが出るはず

terurin replied:
定数に代入はできない

terurin commented on C++/C++/section10.2_class_virtual.md:
```diff
@@ -0,0 +1,138 @@
+>派生クラスではoverrideやvirtualを記述しなくてもオーバーライドは可能である  
+しかし、これらのキーワードがないとシグネチャが異なるときにオーバーライドにならず新しいメソッドを定義したことになる  
+これらのキーワードをつければ記述ミスがあった時などにエラーになってくれてバグを防ぐことができる
+
```
finalに触れれるといいかな

terurin commented on C++/C++/section10.2_class_virtual.md:
```diff
@@ -0,0 +1,138 @@
+しかし、派生クラスのメンバにはアクセスできないので注意
+
+virtualと付ければメソッドは派生クラスにアクセスできる
+
```
static_cast, dynamic_cast

naga-karupi committed:
add section11_cast.md
postscript section10.2_class_virtual.md and column1_const.md

naga-karupi committed:
add section12_lambda
postscript section10.2_class_virtual and section11_cast

naga-karupi committed:
add column2_range_for
postscript section12_lambda

naga-karupi committed:
add colomn1_filesplit
fix column2_range_for

naga-karupi committed:
add colomn2_build_optioin
add makefile_1 but incompleted

naga-karupi committed:
taigen stop!!

naga-karupi committed:
write makefile_1.md

naga-karupi committed:
push test

naga-karupi committed:
add colomn3_rvalue.md, section13.
but section13 has not written yet.

naga-karupi committed:
write section13.1
add name colomn 3

naga-karupi committed:
write section13.2_move.md
fix section9.2_class_constructor.md

naga-karupi committed:
fix sectioin13.2

naga-karupi committed:
add section14.1_smart_pointer.md

naga-karupi committed:
add section14.2_unique_ptr.md
fix section14.1

naga-karupi committed:
add section14.3_shared_ptr_weak_ptr.md

naga-karupi committed:
add section0 in cpp

naga-karupi committed:
rename make

naga-karupi committed:
found std::plus() and improve sample code error

naga-karupi committed:
need {}

naga-karupi committed:
fix typo

naga-karupi committed:
add section 0 ros install

naga-karupi committed:
no use apt-get

naga-karupi committed:
section0-python

naga-karupi committed:
Merge branch 'naga/c_plusplus' of https://github.com/TNCTRobocon/manuals_markdown into naga/c_plusplus

naga-karupi committed:
add md about cmake introduction and simple build

naga-karupi committed:
fix

naga-karupi commented:
mainブランチにしてほしいという要望が出たので検討よろしくおねがいします

terurin commented:
conflictしている分を解決してください。

terurin commented on C++/C++/column2_range_for.md:
````diff
@@ -0,0 +1,111 @@
+    ...
+}
+```
+
````
for文の中で要素の追加・削除を行ってはいけないことを述べておくこと。

terurin commented on C++/C++/section11_cast.md:
```diff
@@ -0,0 +1,190 @@
+判断する手間が増えるので重い処理になるので何回も使わないのがベスト
+
+安全であるなら`static_cast`でもまったく問題ない
+
```
もしshared_ptrにふれるなら、次のキャストを知っておくと良い。

* [static_pointer_cast](https://cpprefjp.github.io/reference/memory/shared_ptr/static_pointer_cast.html)
* [dynamic_pointer_cast](https://cpprefjp.github.io/reference/memory/shared_ptr/dynamic_pointer_cast.html)

terurin commented on C++/C++/section14.1_smart_pointer1.md:
```diff
@@ -0,0 +1,195 @@
+## auto_ptr
+
+※C++11からは非推奨
+
```
(pic32の開発でもしない限り)c++11より前のコードを触れることはなく、読者に余計な負担をかけることになるのでauto_ptrの記述は不要。

terurin replied:
特別な理由がないならauto_ptrは消してしてください。

terurin commented on C++/C++/section14.3_shared_ptr_weak_ptr.md:
```diff
@@ -0,0 +1,178 @@
+### そもそもなんで必要なのか
+
+`shared_ptr`は性質上循環参照を起こす可能性がある
+
```
循環参照とは

terurin commented on C++/C++/section14.3_shared_ptr_weak_ptr.md:
```diff
@@ -0,0 +1,178 @@
+
+こんな感じで書ける
+
+危険なアクセスをする気がしたら`weak_ptr`を使おう
```
weak_ptrは所有権を主張できないので考えて使うように書く

terurin commented on C++/C++/section14.3_shared_ptr_weak_ptr.md:
```diff
@@ -0,0 +1,178 @@
+こんな感じで書ける
+
+危険なアクセスをする気がしたら`weak_ptr`を使おう
+
```
std::make_sharedと型キャストを書くと良い

terurin commented on C++/C++/section3_reference.md:
```diff
@@ -1,145 +1,145 @@
-
-2021/11/30
-writen by 西永
+# 参照
```
変更されていない箇所まで変更されているようにみえるけど、
改行コードの変更でもした?

terurin commented on C++/C++/section9.1_class_introduction.md:
```diff
@@ -0,0 +1,215 @@
+メンバ変数がprivateになると`instance.member`といった方法でアクセスできない変数や関数も出てくる  
+その時にメンバ変数をpublicなメンバ関数を読み書きするためのメンバ関数を**アクセサ**といい、値を取得するアクセッサを**getter**、値を設定するためのものを**setter**という
+
+カプセル化が破壊されること、setterやgetterのような名前だと値を変える意図が伝わりにくくなるので別のメソッド名がないか検討することを勧める
```
強調するといいかも

terurin commented on C_lang/C_lang/colomn2_build_optioin.md:
```diff
@@ -0,0 +1,114 @@
+warningを出せるだけ出してくれる  
+バグの可能性のある部分をあらかじめ警告する  
+付けておいた方がお得
+
```
特殊な事情がない限り、つけること

terurin commented on C_lang/C_lang/colomn2_build_optioin.md:
```diff
@@ -0,0 +1,114 @@
+  コードが生成される可能性があるが思い通りに動く保証はないので推奨されない
+- `Os` : バイナリのサイズを重視した最適化で、`-O2`オプションの一部の機能を制限したものなの
+- `-Og` : デバッグ機能を維持しつつコードの最適化をいい感じに行う
+
```
`-O0`と`-O2`しか基本的に使わない

terurin commented on C_lang/cmake/section0_introduction.md:
```diff
@@ -0,0 +1,150 @@
+となる
+
+これでgitを使おうとすると大変だろうなということが伝わるだろうか
+
```
gitignoreがあるのでgitではなく、純粋に管理しづらいでいいかと

terurin commented on C_lang/cmake/section0_introduction.md:
```diff
@@ -0,0 +1,150 @@
+
+> cd build
+
+> make
```
次のようにビルドすることが多い気がする
```sh
mkdir -p build; cd build
cmake ..&&make -j &&./<your program>
```

terurin commented on C_lang/cmake/section1_simple_build.md:
```diff
@@ -0,0 +1,188 @@
+
+正直`hello.cpp`というファイルを追加しただけなので解説することはないとは思う  
+少なくとも同じファイル内であればこのように追加していけばいくらでもファイルのリンクはできる
+
```
subdirectoryに付いて触れられるとよい。
また、ctestを調べておくと良い。

terurin commented on C_lang/make/makefile.md:
````diff
@@ -0,0 +1,165 @@
+~~説明する内容が変数だけじゃないのは内緒~~
+
+```makefile
+TARGET = main
````
`=`にすると保守性に難が生じるので`?=`を用いてオプションを設定すると良い

naga-karupi committed:
Merge branch 'main' of github.com:TNCTRobocon/manuals_markdown into naga/c_plusplus

terurin commented on C++/C++/section14.3_shared_ptr_weak_ptr.md:
```diff
@@ -0,0 +1,178 @@
+# shared_ptrとweaK_ptr
```
weak_ptr

naga-karupi committed:
ファイルの移動

naga-karupi committed:
微修正

naga-karupi committed:
README.mdの更新

naga-karupi committed:
makefile の代入演算子について追記と微修正

naga-karupi committed:
mainブランチのマージ

naga-karupi committed:
auto_ptrの排除

naga-karupi committed:
static_pointer_cast,dynamic_pointer_castの説明の追加

naga-karupi committed:
未定義動作と出力の追記

naga-karupi committed:
注意書きを追加

naga-karupi committed:
コメント追加

naga-karupi committed:
隠ぺいについて例を追加

naga-karupi committed:
cmakeについて追記

naga-karupi committed:
subdirectoryについて追記

naga-karupi committed:
typoの修正
